src/coreclr/vm/jitinterface.cpp

```diff
@@ -10859,41 +10845,70 @@ void CEECodeGenInfo::getHelperFtn(CorInfoHelpFunc    ftnNum,               /* IN
                     {
                         // Cache it for future uses to avoid taking the lock again.
                         hlpFinalTierAddrTable[dynamicFtnNum] = finalTierAddr;
-                        if (pNativeEntrypoint != NULL)
-                        {
-                            pNativeEntrypoint->accessType = IAT_VALUE;
-                            pNativeEntrypoint->addr = finalTierAddr;
-                        }
+                        accessType = IAT_VALUE;
+                        targetAddr = finalTierAddr;
                         goto exit;
                     }
                 }
             }
 
             if (IndirectionAllowedForJitHelper(ftnNum))
             {
+#ifdef FEATURE_PORTABLE_ENTRYPOINTS
+                accessType = IAT_VALUE;
+                targetAddr = (LPVOID)helperMD->GetPortableEntryPoint();
+#else // !FEATURE_PORTABLE_ENTRYPOINTS
                 Precode* pPrecode = helperMD->GetPrecode();
                 _ASSERTE(pPrecode->GetType() == PRECODE_FIXUP);
-                if (pNativeEntrypoint != NULL)
-                {
-                    pNativeEntrypoint->accessType = IAT_PVALUE;
-                    pNativeEntrypoint->addr = ((FixupPrecode*)pPrecode)->GetTargetSlot();
-                }
+                accessType = IAT_PVALUE;
+                targetAddr = ((FixupPrecode*)pPrecode)->GetTargetSlot();
+#endif // FEATURE_PORTABLE_ENTRYPOINTS
                 goto exit;
             }
         }
 
-        pfnHelper = LoadDynamicJitHelper((DynamicCorInfoHelpFunc)dynamicFtnNum).pfnHelper;
+        pfnHelper = LoadDynamicJitHelper(dynamicFtnNum);
```

Comment:
This still does not look right. LoadDynamicJitHelper can return IL-based helper or dynamically assigned native helper. The dynamically assigned helper still needs wrapping by a portable entrypoint. The helpers backed by managed methods are the only ones where we can skip the wrapping.

Reply:
I think that the maze of the different ifdefs is impossible to deal with well using interleaved ifdefs. I have pushed a commit into your PR with what I think it should look like. (I have not tested - please beat it into shape.)

I do not really like that `LoadDynamicJitHelper` can return either portable entrypoint or actual code. Refactoring that contract can be done separately.